### PR TITLE
incorporation of PWFCF-JCorr analysis

### PR DESCRIPTION
**Pull request by dongjokim**

### CODE REVIEW

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]
Hi Don Jo! Welcome to O2!!

You did right in creating JCORRAN  sub-directory but please, use JCorran instead or something perhaps more explicit
Then, you have to follow the standard layout under this sub-directory
You have to have under it
- Core: you put there general classes, if any, that are used by the tasks in the scope of JCorran
- DataModel: you put there the declaration of the tables tables and constants/values/structures which will be internally used by the tasks in the scope of JCorran
- TableProducer: you put there the task within JCorran which produce internal tables or derived data which will be used by the tasks in the scope of JCorran
- Tasks: finally you put here the JCorran tasks

So, if I interpreted correctly your code
- AliJFFlucAnalysis.h/cxx, AliJFFlucAnalysis.h/cxx, AliJHistManager.h/cxx, and JCorrAnLinkDef.h should go to JCorran/Core. Please, get rid of the "Ali" prefixes
- AliJO2Catalyst.cxx should go to JCorran/TableProducer. Please, get rid of the "Ali" and "O2" prefixes. Please, don't use 'class' for the task, use 'struct' 
- AliJO2Analysis.cxx. should go to JCorran/Tasks. Please, get rid of the "Ali" and "O2" prefixes. Please, don't use 'class' for the task, use 'struct'. You need to re-engineer this. You are not allowed to replicate the data you received. This penalizes a lot the execution and the required memory. You have to use the original table or the derived tables but not copy them to your own data structures. That's not the aim of O2
- AliJO2Catalyst.h should go to JCorran/DataModel. Pleae get rid of the "Ali" and "O2" prefixes

You will need a CMakeLists.txt at the Jcorran level and additional ones at its sub-directories levels that have source code
Please, remove all "Ali" prefixes and don't put "O2". We don't need this anymore

Please, remove all code lines that are left commented (not executed). Your code will be in production so only what is executed should be in the source file
Of course, comments to make your code more readable are more than welcome

For the time being I would suggest you to put the PR in draft until you address all the above

**Comment by jgrosseo** [comment]
Thanks for the extensive review @victor-gonzalez 
I have changed the PR to draft to reduce the load on the CI. Once things have converged you can just put it back to "ready for review".

**Comment by jaelpark** [comment]
Hi @victor-gonzalez!
Thanks a lot for your review. I've made the changes according to the requirements. @dongjokim should have a look too.
-Jasper

**Comment by dongjokim** [comment]
thanks, victor for your comments, and jasper for taking care of the rest. I think we are ready for another look from Victor.
Cheers,
DJ

**Comment by victor-gonzalez** [comment]
I will review it being in draft also so following Jan Fiete's suggestion please put it in draft to not overload CI

File: PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx
Comment (victor-gonzalez) [inline PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:115]: I guess at this point you are creating a copy of the data in the `ParticleTrack` table to your internal structure which will be used in your `UserExec` routine
You have to have in mind that this is not the aim of the O2 framework. This will probably impact the task execution time as well as its memory requirements which will penalize the execution on grid
I would suggest to adapt the `UserExec` to work directly with the tables if feasible

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]

File: PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx
Comment (jgrosseo) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:115]: Yes, this is indeed an important point. @jaelpark we can also discuss about this with a few examples how one can do it without making copies. E.g. in `correlations.cxx` are a few examples...

Comment (jaelpark) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:115]: In fact, the new version doesn't copy any tables or particle lists. Here ```Tracks``` is merely an iterator interface, that the correlations class ```JFFlucAnalysis``` in Core can use to access the original particle table without O2 dependencies. There's no internal structure anymore, but there's a wrapper around the native ```ParticleTrack``` iterator.

File: PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx
Comment (jgrosseo) [inline PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:46]: Is this not copying the actual values, @jaelpark ?

Comment (jaelpark) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:46]: @jgrosseo Yeah it puts them into a ```PtEtaPhiEVector```, but at least not for the entire list at once, that I understood was not allowed from Victor's comments. But if ```PtEtaPhiEVector``` (and such) are forbidden, then I would do some heavier changes @dongjokim.

Comment (jgrosseo) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:46]: Ideally everything is zero copy which can be achieved with some templated functions or classes but of course one cannot replay a full framework which copies things. I propose that we briefly discuss this tomorrow when we meet anyway before you go ahead and do the changes. How does that sound?

Comment (jaelpark) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx:46]: Yep, okay thanks. Anyways I'll probably get back to this later, have to prepare for PF.

File: PWGCF/JCorran/Core/JCORRANLinkDef.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h:17]: hello, is it really necessary to change all names of our classes? We have to waste our time for nothing. Can you explain why this rule was in place?

Comment (jgrosseo) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h:17]: I understand this is a little work but we do not have AliPhysics anymore and the Ali refers to this. Also for O2 there should not be O2 in the class names.
Have you already tried to use Visual Studio Code for it? It has a very good "replace in multiple files" feature.

Comment (dongjokim) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h:17]: The codes don't matter if they are in the AliPhysics or O2, they are just classes that are used on other occasions, like toy mc etc. If this is the rule enforced for o2, we can change them. We were just forced to add the prefix.

File: PWGCF/JCorran/Core/JFFlucAnalysis.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Core/JFFlucAnalysis.cxx:480]: This is not what we want, the methods are different and the observables are not exactly identical for many occasions.

File: PWGCF/JCorran/Core/JFFlucAnalysis.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JFFlucAnalysis.h:137]: as long as they are in our own class, we think they are just fine.

File: PWGCF/JCorran/Core/JHistManager.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Core/JHistManager.cxx:251]: I don't understand why, If it is with LOGF, this class needs a full ALICE framework. If it is not the case, can you point me the lib of this LOGF?

**Comment by dongjokim** [comment]
As for Ali in JHistoManager, they are just internal namings. I hope the rule is for the actual naming of the code.
Other than this, I think we are ready for merging.

File: PWGCF/JCorran/Core/JCORRANLinkDef.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h]: done!

File: PWGCF/JCorran/Core/JFFlucAnalysis.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JFFlucAnalysis.h]: as long as they are in our own class, we think they are just fine.

File: PWGCF/JCorran/Core/JHistManager.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JHistManager.h]: as long as they are in our own class, we think they are just fine.

File: PWGCF/JCorran/Core/JCORRANLinkDef.h
Comment (jgrosseo) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h]: @dongjokim I don't understand your comment about where the code is. 
The rule which has been agreed and followed with all code in O2Physics up to now is that we do not keep the `Ali` prefix. E.g. AliFemto had to do a major work. I think the same should apply to your code. You write "done" but there are still many classes with Ali. What is your worry here, the additional work?
We can also involve @ddobrigk if needed here as analysis coordinator.

I anyway still have to see with @jaelpark how to refactor the code which copies the track values. Unfortunately, we were not able to discuss this last week due to reasons outside of our control...

**Comment by jgrosseo** [comment]
Your rebase may have failed because you have now commits from @pbuehler here which does not seem right...

File: PWGCF/JCorran/Core/JCORRANLinkDef.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h]: ok, thanks. I meant for the naming inside the code as well as the code naming. It looks like this can confuse people. Now I have changed the inner class's naming as well.

File: PWGCF/JCorran/Core/JFFlucAnalysis.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JFFlucAnalysis.h]: done as suggested.

File: PWGCF/JCorran/Core/JHistManager.h
Comment (dongjokim) [reply PWGCF/JCorran/Core/JHistManager.h]: done

File: PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx]: done

File: PWGCF/JCorran/Core/JCORRANLinkDef.h
Comment (jgrosseo) [reply PWGCF/JCorran/Core/JCORRANLinkDef.h]: Great - thanks! I will see with @jaelpark about the copying and then get back to you.

File: PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx]: we will change to JTracks and JTracksBase

File: PWGCF/JCorran/Core/JHistManager.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Core/JHistManager.cxx]: will change to LOGF

File: PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Tasks/jflucAnalysisTask.cxx]: it does 't exist any more

File: PWGCF/JCorran/Core/JHistManager.cxx
Comment (dongjokim) [reply PWGCF/JCorran/Core/JHistManager.cxx]: done :https://github.com/AliceO2Group/O2Physics/pull/1244/commits/63cbc533c74255711f1b1c50c78daa804659a2e2

**Comment by vkucera** [comment]
Hi @dongjokim , sorry to jump in but please give your PR a meaningful title. The current one does not say anything at all about the code changes.

**Comment by victor-gonzalez** [comment]
You need to rebase your changes
And as Vit suggested it would be good to set a proper name for your PR
What about "incorporation of JCorr analysis"?

**Comment by dongjokim** [comment]
> 

yes, it is changed now.

File: PWGCF/JCorran/DataModel/JCatalyst.h
Comment (victor-gonzalez) [inline PWGCF/JCorran/DataModel/JCatalyst.h:26]: Please, comment your column declaration for derived tables so they can appear in the documentation

Comment (victor-gonzalez) [reply PWGCF/JCorran/DataModel/JCatalyst.h:26]: For guidance check `Common/DataModel/Centrality.h`

File: PWGCF/JCorran/DataModel/JCatalyst.h
Comment (victor-gonzalez) [inline PWGCF/JCorran/DataModel/JCatalyst.h:33]: Same as above

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]
Thanks for implementing all our suggestions!
There is only one final comment regarding documenting the column declaration
We will approve it once ready

File: PWGCF/JCorran/DataModel/JCatalyst.h
Comment (jaelpark) [reply PWGCF/JCorran/DataModel/JCatalyst.h]: Ok, added!

**Review by victor-gonzalez — APPROVED** [review]
Thanks!

**Comment by jaelpark** [comment]
Thanks for the approval, Victor. I still had to remove an unused class member that apparently was failing the build test.

**Review by victor-gonzalez — APPROVED** [review]

**Pull request merged**
